## 由一个报错引起的对原型、赋值的思考

### 前言：一个报错

由于这周的周会是有 Shiina 分享关于 JavaScript 原型和继承的深入了解

为了更好地理解成员分享的知识，我重新复习了一下相关知识，稍加折腾便遇到了以前从未动手尝试过的问题

首先，我折腾了这样一段代码：

```javascript
const Person = function() {
    return {
        name: "coco"
    }
}

Person.prototype.name = "Uni"
Person.prototype.sayName = function() {
    console.log(this.name);
}

const person = new Person()
person.sayName();
```

![Jv17E4.png](https://s1.ax1x.com/2020/05/02/Jv17E4.png)

这段代码会造成以上错误，错误提示很清楚，知道错误的点也很轻松，但是怎么通过这个报错吧原型的基本知识讲清楚是本文的目的



### 讨论原型前我们先来看看构造函数吧

众所周知，构造函数可以用来创建特定类型的对象。首先我们先来构建一个构造函数

```javascript
const Person = function(name, age, stack) {
    this.name = name
    this.age = age
    this.stack = stack
}
```

通过 `new` 关键字进行实例化对象，所以我们无法将箭头函数当做一个构造函数使用，要知道箭头函数是不能使用 `new` 关键字进行调用，而且箭头函数没有 `this` ，也没有 `arguments`



**new调用构造器的时候，都发生了什么？**

在我们使用 `new` 关键字调用一个函数的时候，这个函数被称作构造器或者构造函数。在 `new` 一个构造函数的时候，构造函数会默认、隐式地 `return ` 一个对象，这个对象我们称作 `this`

```javascript
const Person = function(name, age, stack) {
    this.name = name
    this.age = age
    this.stack = stack
    this.show = function() {
        console.log(this);
    }
}

const person = new Person("Uni", 19, "react");
console.log(person)   
person.show()
```

![JvtW8J.png](https://s1.ax1x.com/2020/05/02/JvtW8J.png)

我们通过结果可以看到结论是对的。

**但如果此时我们手动去 `return` 一个对象，会发生什么？**

这种主动去 `return` 一个非 `this` 对象，我们称作显示返回。

```javascript
const Person = function(name, age, stack) {
    this.name = name
    this.age = age
    this.stack = stack
    this.show = function() {
        console.log(this);
    }
    return {
        name: "tea",
        age: "20",
        lover: "Uni"
    }
}

const person = new Person("Uni", 19, "react");
console.log(person)
```

![JvNZMn.png](https://s1.ax1x.com/2020/05/02/JvNZMn.png)

实例化对象 `person` 变为了显示返回的对象，这个是显而易见的。



### 再来看看我们的ECMAScript的本质：原型

要知道，我们 `JavaScript` 的面向对象系统是基于原型编写的，参考 `Self` 和 `Smalltalk` 这两门原型语言

所以说，我们的ES6新规范 `class` 只不过是一个对 `原型及原型继承` 进行了封装的语法糖

当我们创建了一个函数时，引擎会通过一系列的规则为我们绑定上我们的 `prototype` 属性，而这个属性指向的是一个对象，这个对象我们便叫做 `原型对象 `

```javascript
const Person = () => {}
const Human = function() {}

console.log(Person.prototype)
console.log(Human.prototype)
```

![JvR1Y9.png](https://s1.ax1x.com/2020/05/02/JvR1Y9.png)

我们可以观察到，箭头函数并不会绑定上 `prototype` 属性



原型对象存在的意义便是能够共享资源，它所开辟的内存地址所存放的方法、属性都是能够被相关实例化对象调用的。

既然构造函数上有一个属性能够指向它的原型对象(即 prototype属性)，那原型对象上有没有属性能够指向它的构造函数呢？

**我们的每个原型对象上，都会隐式挂载一个叫做 `constructor` 的属性，这个属性便会指向此原型对象的构造函数**

好了我们来验证一下刚刚说的两点吧：

- constructor的存在
- 原型对象能够给该构造函数实例化的对象提供公共成员变量

```javascript
const Person = function() {}
const person = new Person()

// constructor的存在
console.log(Person.prototype.constructor === Person)  // true

// 原型对象能够给该构造函数实例化的对象提供公共成员变量
console.log(person.constructor === Person.prototype.constructor)   // true
```

我们实例化的对象，本身是没有 `constructor` 的， 在调用没有定义的方法时，它会调用原型链进行查找，看看原型对象上有没有公共的方法

那它是如何查找原型对象的呢？也就是说原型链是什么呢？

**我们刚刚有提到的是，在使用 `new` 通过构造函数实例化对象时，构造函数会返回一个 `this` 对象，而这个对象上会隐式地绑定上一个叫做 `__proto__` 的属性，这个也是产生这次报错的根本原因**

每个普通的由构造函数实例化的对象上都会绑定一个`__proto__` ，这个属性指向的便是它的**构造函数的原型对象**，所以便能够访问到公共资源。



### 问题迎刃而解

所以我们造成此次报错的原因很简单，因为我们的构造函数显示返回了一个对象，而这个对象上是不会绑定`__proto__` 这个属性，自然也访问不到相应的原型对象



 在我的名为 [一次连续赋值体验](http://cocotea.life/index.php/archives/25/) 的这篇文章中有写道，引用类型下调用不存在的变量并不会产生“未声明该变量”的报错，而是会隐式地赋予一个 `undefined`。而对 `undefined` 进行调用自然而然会报出“该变量不是函数”这样的错误